### ISSUE TITLE
Subo build docker toolchain for Rust painfully slow

### ISSUE THREAD

**Issue by flaki** [suborbital/subo#181]
Building a simple `helloworld` runnable might take up to 30s while using the built-in docker toolchain on a recent (4-core Intel Whiskey Lake mobile CPU) laptop running Ubuntu 20.04. As I can see the majority of this time is actually spent in the toolchain bootstrap phase, on _"Updating crates.io index"_.

![image](https://user-images.githubusercontent.com/2089432/151631944-64575857-c6cc-4055-9005-b03fd911f742.png)

This is supported by the fact that the toolchain downloads ~40MB during this phase, so the process doesn't seem to be CPU-bound, but rather at least partially network-bound. As a first step we should explore whether we could preload or cache the updated crates.io registry information for subsequent toolchain executions (the pre-initialization/cold start problem all over again!), because not only this makes the builds excruciatingly slow, but will also make them wasteful of network bandwith.

**Comment by cohix** [comment]
Yep, this has been a thorn in my side since the beginning. I did make a change a while ago to build all {lang} runnables with a single container job rather than one job per runnable, so at least if you have 5 Rust runnables it only does the Crates.io work once.

That said, it's super annoying and I'd love to figure out how to properly cache that data on the host.

**Comment by cohix** [comment]
Related: https://github.com/suborbital/subo/issues/12

**Comment by flaki** [comment]
I think in the case of Rust perhaps the easiest fix would be to mount a local volume into the build container that would persist this data. We could either mount it in `$HOME`, or use the [`CARGO_HOME`](https://doc.rust-lang.org/cargo/reference/config.html) environmental variable to put `cargo`'s home folder into the mounted volume's path.

Cargo puts the index in `~/.cargo/registry/index`, which itself is a simple [git repository](https://github.com/rust-lang/crates.io-index) that tracks the published crates, so as long as it is preserved across toolchain invocations, cargo should just update it through a simple pull and we wouldn't need to worry about it getting out of sync or stuck.

If we want to do more caching (e.g. proper build cache for runnables) we could preserve the build directories for those as well, although in that case a "cache clean" command will probably come useful, should something go awry.

**Comment by rnpridgeon** [comment]
https://github.com/suborbital/subo/pull/245
